### ISSUE TITLE
Union types don't work as React props type for component classes

### ISSUE THREAD

**Issue by poeschko** [facebook/flow#3809]
This code:
```
import React from 'react';

type Props = {prop: "foo"} | {prop: "bar"};

class Comp extends React.Component {
  props: Props;
  render() {
    return <span>this.props.prop</span>;
  }
}

<Comp prop="foo" />
```
gives this error:
```
<Comp prop="foo" />
^ props of React element `Comp`. Could not decide which case to select
props: Props;
       ^ union type
```

## Additional investigation & notes

Using the union type by itself works as expected:
```
const p1: Props = {prop: "foo"};
const p2: Props = {prop: "bar"};
// Error (as expected): const p3: Props = {prop: "baz"};
```

If I remove the union and make it just
```
type Props = {prop: "foo"};
```
then
```
<Comp prop="foo" />
```
also works as expected.

It does not make a difference whether I make the `prop` optional or not (using `prop?: ...`).

Also, this seems to be specific to class-based components: Using a functional component
```
function Comp(props: Props) {
  return <span>{props.prop}</span>;
}
```
works even with a union type for `Props`.

I tried this on https://flow.org/try with all released versions v0.42.0 through v0.44.1 and master.

**Comment by poeschko** [comment]
Possibly related: issue #2508 

Indeed (inspired by [jdan's comment](https://github.com/facebook/flow/issues/2508#issuecomment-271975166)), when I add `defaultProps` as in
```
class Comp extends React.Component {
  props: Props;
  static defaultProps = {prop: "foo"};
  render() {
    return <span>this.props.prop</span>;
  }
}
```
things work as expected (no "Could not decide which case to select" error).

**Comment by asolove** [comment]
This problem with inferring prop types (especially with default prop types) is now resolved in Flow 0.53, which forces `Component`s to declare the prop types. With a tiny bit of syntax change, [the original example now type-checks](https://flow.org/try/#0JYWwDg9gTgLgBAJQKYEMDG8BmUIjgcilQ3wG4AocmATzCTgAUcwBnOAXjgG8xmAuOACJMECIIC+cAD7deEMAMEAjFFAkVyaADYoWbAMK4wcJAA8YSAHYATNsnQwAdIfARLVmAB4m8lgBo4ACoAPm5yODg5VgEfVgoIohskKAAKAEowiISkGABXKEs4TxYwFEtgmAALYBZHKNqozwB6ErLg+Lhxci7yTxdjKPZhUUE4JuCgA).

I believe this issue can be closed (/cc @calebmer)

**Comment by sam-rad** [comment]
This doesn't seem to work with `function` components. See [the example:](https://flow.org/try/#0JYWwDg9gTgLgBAJQKYEMDG8BmUIjgcilQ3wG4AocmATzCTgAUcwBnOAXjgG85MAbFAHMAXHABGECH1QA7OAF84AH25wZKEElFwWMKMBmCFFcpgCuMjMAhyAwrjAAKMMxaimEVgEpu5OHCIYMyg5AB4WMBQZAD4XTxYAOn4hUIB6CKjoinlKUPtwODiwdgAiTEkSuFTooA)

```js
import React from 'react';

type Props = { flag: boolean } | { name:  string };

function Comp(props: Props) {
  return <span>props.flag</span>;
}

<Comp prop="foo" />
```
Which throws this error:
```
<Comp prop="foo" />
    ^ Cannot create `Comp` element because: Either property `flag` is missing in props [1] but exists in object type [2]. Or property `name` is missing in props [1] but exists in object type [3].

    References:

    9: <Comp prop="foo" />
       ^ [1]

    3: type Props = { flag: boolean } | { name:  string };
                    ^ [2]

    3: type Props = { flag: boolean } | { name:  string };
                                        ^ [3]
```
